src/service/viewer-impl.js

```diff
@@ -392,6 +409,84 @@ export class Viewer {
   }
 
   /**
+   * Returns the "referrer" URL that can be optionally customized by
+   * the viewer.
+   * @return {string}
+   */
+  getReferrerUrl() {
+    // If not embedded in a IFrame always return the document's referrer.
+    if (!this.isEmbedded()) {
+      return this.win.document.referrer;
+    }
+
+    // Viewer override, but only for whitelisted viewers.
+    if (this.params_['referrer'] && this.isTrustedViewer()) {
+      return this.params_['referrer'];
+    }
+
+    // Document's referrer.
+    if (this.win.document.referrer) {
+      return this.win.document.referrer;
+    }
+
+    // Ancestor in the embedded case.
+    const ancestorOrigins = this.win.location.ancestorOrigins;
+    if (ancestorOrigins && ancestorOrigins.length > 0) {
+      return ancestorOrigins[0];
+    }
+
+    // No referrer found.
+    return '';
+  }
+
+  /**
+   * Whether the viewer has been whitelisted for more sensitive operations
+   * such as customizing referrer.
+   * @return {boolean}
+   */
+  isTrustedViewer() {
+    return this.isTrustedViewer_;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  calcTrustedViewer_() {
+    // This only applies to the documents embedded as iframes.
+    if (!this.isEmbedded_) {
+      return false;
+    }
+
+    // Ancestors when available take precedence. This is the main API used
+    // for this determination. Fallback is only done when this API is not
+    // supported by the browser.
+    const ancestorOrigins = this.win.location.ancestorOrigins;
+    if (ancestorOrigins) {
+      return (ancestorOrigins.length > 0 &&
+          this.isTrustedViewerOrigin_(ancestorOrigins[0]));
+    }
+
+    // Fallback to the referrer if ancestorOrigins is not available.
+    return this.isTrustedViewerOrigin_(this.win.document.referrer);
+  }
+
+  /**
+   * @param {string} urlString
+   * @return {boolean}
+   * @private
+   */
+  isTrustedViewerOrigin_(urlString) {
+    const url = parseUrl(urlString);
+    for (let i = 0; i < TRUSTED_VIEWER_HOSTS.length; i++) {
```

Comment:
I think at the latest we decided not to use it, but I see that we are already using it elsewhere. Replaced with `some`.